README.md
# Swap-Mukham WIP

## Changes
- added parellel execution utilizing concurrent.futures
- added test preview button
- added video frame slider
- added collect face from preview
- added frame selection to replace trim feature
- converted every model to onnx format
- editable onnx execution providers at ``utils/device``
- removed moviepy, insightface dependency
- removed pytorch dependency for cpu users
- command line arg ``--prefer_text_widget`` to replace target video widget with text
- more face-upscaler support (restoreformer, codeformer, gfpgan, gpen)
- added face-upscaler opacity slider
- nsfw-detector now only checks random 100 frames
- added custom foreground painting
- ui changes (gradio 3.40)
- added swap-iteration (may increase face likeliness)
- added date-time option for output name

  
## Install WIP branch
### CPU Install
````
git clone -b new-wip --single-branch https://github.com/harisreedhar/Swap-Mukham.git
cd Swap-Mukham
conda create -n swapmukham python=3.10 -y
conda activate swapmukham
pip install -r requirements_cpu.txt

python app.py --cpu --prefer_text_widget
````
### GPU Install (CUDA)
````
git clone -b new-wip --single-branch https://github.com/harisreedhar/Swap-Mukham.git
cd Swap-Mukham
conda create -n swapmukham python=3.10 -y
conda activate swapmukham
conda install pytorch==2.0.0 torchvision==0.15.0 torchaudio==2.0.0 pytorch-cuda=11.8 -c pytorch -c nvidia
pip install -r requirements.txt

python app.py --prefer_text_widget
````

## Download and place these models under ``assets/pretrained_models``

- [inswapper_128.onnx](https://huggingface.co/deepinsight/inswapper/resolve/main/inswapper_128.onnx)
- [faceparser.onnx](https://huggingface.co/bluefoxcreation/Face_parsing_onnx/resolve/main/faceparser.onnx)
- [det_10g.onnx](https://huggingface.co/bluefoxcreation/insightface-retinaface-arcface-model/resolve/main/det_10g.onnx)
- [w600k_r50.onnx](https://huggingface.co/bluefoxcreation/insightface-retinaface-arcface-model/resolve/main/w600k_r50.onnx)

#### Face upscalers ``optional``
- [codeformer.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/codeformer.onnx)
- [GFPGANv1.2.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/GFPGANv1.2.onnx)
- [GFPGANv1.3.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/GFPGANv1.3.onnx)
- [GFPGANv1.4.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/GFPGANv1.4.onnx)
- [GPEN-BFR-256.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/GPEN-BFR-256.onnx)
- [GPEN-BFR-512.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/GPEN-BFR-512.onnx)
- [restoreformer.onnx](https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/download/Models/restoreformer.onnx)


history

# 2023-08-28 21:32:06.191874
+Как мне сделать чтобы если был --fp16 то INSWAPPER_PATH = "./assets/pretrained_models/inswapper_128.onnx" иначе INSWAPPER_PATH = "./assets/pretrained_models/inswapper_128.fp16.onnx"

# 2023-08-28 21:33:08.628496
+INSWAPPER_PATH = (
+    "./assets/pretrained_models/inswapper_128.onnx"
+    if user_args.fp16
+    else "./assets/pretrained_models/inswapper_128.fp16.onnx"
+)

# 2023-08-28 21:33:48.693033
+user_args" не определен

# 2023-08-28 21:36:22.574514
+но user_args указываетсчя в app.py


face_upscaler.py
import os
import cv2
import default_paths as dp
from upscaler.GPEN import GPEN
from upscaler.GFPGAN import GFPGAN
from upscaler.codeformer import CodeFormer
from upscaler.restoreformer import RestoreFormer

def gfpgan_runner(img, model):
    img = model.enhance(img)
    return img


def codeformer_runner(img, model):
    img = model.enhance(img, w=0.9)
    return img


def gpen_runner(img, model):
    img = model.enhance(img)
    return img


def restoreformer_runner(img, model):
    img = model.enhance(img)
    return img


supported_upscalers = {
    "CodeFormer": (dp.CODEFORMER_PATH, codeformer_runner),
    "GFPGANv1.4": (dp.GFPGAN_V14_PATH, gfpgan_runner),
    "GFPGANv1.3": (dp.GFPGAN_V13_PATH, gfpgan_runner),
    "GFPGANv1.2": (dp.GFPGAN_V12_PATH, gfpgan_runner),
    "GPEN-BFR-512": (dp.GPEN_BFR_512_PATH, gpen_runner),
    "GPEN-BFR-256": (dp.GPEN_BFR_256_PATH, gpen_runner),
    "RestoreFormer": (dp.RESTOREFORMER_PATH, gpen_runner),
}

cv2_upscalers = ["LANCZOS4", "CUBIC", "NEAREST"]

def get_available_upscalers_names():
    available = []
    for name, data in supported_upscalers.items():
        if os.path.exists(data[0]):
            available.append(name)
    return available


def load_face_upscaler(name='GFPGAN', provider=["CPUExecutionProvider"], session_options=None):
    assert name in get_available_upscalers_names() + cv2_upscalers, f"Face upscaler {name} unavailable."
    if name in supported_upscalers.keys():
        model_path, model_runner = supported_upscalers.get(name)
    if name == 'CodeFormer':
        model = CodeFormer(model_path=model_path, provider=provider, session_options=session_options)
    elif name.startswith('GFPGAN'):
        model = GFPGAN(model_path=model_path, provider=provider, session_options=session_options)
    elif name.startswith('GPEN'):
        model = GPEN(model_path=model_path, provider=provider, session_options=session_options)
    elif name == "RestoreFormer":
        model = RestoreFormer(model_path=model_path, provider=provider, session_options=session_options)
    elif name == 'LANCZOS4':
        model = None
        model_runner = lambda img, _: cv2.resize(img, (512,512), interpolation=cv2.INTER_LANCZOS4)
    elif name == 'CUBIC':
        model = None
        model_runner = lambda img, _: cv2.resize(img, (512,512), interpolation=cv2.INTER_CUBIC)
    elif name == 'NEAREST':
        model = None
        model_runner = lambda img, _: cv2.resize(img, (512,512), interpolation=cv2.INTER_NEAREST)
    else:
        model = None
    return (model, model_runner)